[MRG] Fix of Spectral embedding implementation

jmargeta opened the pull request:
#### Reference Issue
Fixes #8129
Continuation of and closes #8217 by @devanshdalal 


#### What does this implement/fix? Explain your changes.
This fixes computation of the spectral embedding
(normalization of the spectrum by division instead of multiplication).

The first eigenvector of the graph Laplacian is now constant (property of fully connected graphs).

These changes add tests for untested codebase.
*NOTE:* Parts depending on pyamg are not automatically tested on Travis
Testing of all backends would require to install optional package pyamg.

Replaces AMG solver test with a more complete test
     - when the pyamg dependency is installed, the original fails also on master (untouched for 5 years)
     - the new test based on scikit-learn toy image segmentation example instead
       compares results of all three solvers to be the same


#### Any other comments?

The examples are mostly visually similar. With two exceptions:

 - the quality of the results for discretized label assignment method in examples/cluster/plot_face_segmentation.py demo qualitatively improves ![discretize](https://user-images.githubusercontent.com/1239359/26930776-c4efb3f4-4c5e-11e7-8f9c-d7551d304339.png).
In this example a new seed was set to avoid failed plotting for "discretize" option due to empty clusters (matplotlib's contour fail)

 - in examples/manifold/plot_compare_methods.py, for manifold comparison for SpectralEmbedding looses its spread. A more spread out distribution is reported for similar example in the original papers.  ![manifold_method_comparison](https://user-images.githubusercontent.com/1239359/26930629-42707f4e-4c5e-11e7-834a-f43fdbe6f9cc.png).

Commit:
spectral_embedding bug-fix and suggestions

Commit:
spectral_embedding bug-fix and suggestions

Commit:
spectral_embedding bug-fix and suggestions

GaelVaroquaux committed:
Fix spectral clustering

To follow fix in spectral embedding

jmargeta committed:
Fix style of spectral_embedding

jmargeta committed:
Add spectral clustering tests

Replaces AMG solver test with a more complete test
 - NOTE: paths depending on pyamg are not automatically tested on Travis
 - when the pyamg dependency is installed, the original fails also on master
   probably for a very long time
 - test based on scikit-learn toy image segmentation example instead
   comparing results of all three solvers to be the same

jmargeta committed:
Add test of properties of graph Laplacian spectrum

For fully connected graphs the first eigenvector should be constant
while the others not.

jmargeta committed:
Fix spectral clustering

First eigenvector should be kept for clustering.
It is only constant for fully connected graphs.

Reverts 95d8111464fd5a8cd8b19877b1d3c45ee968a344

jmargeta committed:
Change seed for face segmentation example

Fix a seed so that the example would not fail in plotting.
"discretize" might return empty clusters causing contour to fail

jmargeta committed:
Revert to the use of shortened diffusion maps

jmargeta commented:
What do you think @GaelVaroquaux @glemaitre @massich ?

agramfort commented:
LGTM

glemaitre commented:
@agramfort Do you have any thoughts regarding the visible difference on the S-curve dataset.
Originally, we had this: http://scikit-learn.org/stable/auto_examples/manifold/plot_compare_methods.html#sphx-glr-auto-examples-manifold-plot-compare-methods-py

The embedding is more compact now.

jmargeta commented:
This is how it looks for the other examples:
![digits](https://user-images.githubusercontent.com/1239359/26936280-8070d540-4c6e-11e7-9438-4fb75aecef0f.png)
![plot_manifold_sphere](https://user-images.githubusercontent.com/1239359/26936396-e3f009a6-4c6e-11e7-8e48-945f41ef5810.png)

agramfort commented:
indeed it's not as nice ... hum

vene commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -258,7 +258,26 @@ def test_spectral_embedding_unnormalized():
+                                       random_state=seed)
+
+        assert_array_almost_equal(np.diff(embedding[:, 0]), 0)
+        assert_array_equal(np.abs(np.diff(embedding[:, 1])) > 1e-3, True)
```
The second test asserts that no two consecutive elements are close, which seems stricter than what is needed?

jmargeta replied:
Good point @vene  Thanks for your input. 
All I want is a test asserting that the elements are different (not constant).
The condition can be probably relaxed a bit with something like `assert_array_equal(np.std(embedding[:, 1]) > 1e-3, True)`.
Is this any better? What would you suggest instead of the current test?

vene replied:
there are two goals here:

1. make the assertions emantically as close to what we really mean
2. make the assertions understandable by a new coder one year from now

I think you could use the std of the vector for both tests: for the first test the std should be nearly 0, for the second it should be nonzero. but add a small comment next to each?

A stronger idea for the first assertion: don't we mathematically know exactly what value the first component needs to have? (maybe at least when normed?) I'm not sure off the top of my head

GaelVaroquaux commented:
Spectral embedding does look broken.

glemaitre commented:
I spent a little bit of time checking the code and I think that the only tricky part is actually in the computation of the laplacian. I did a naive/textbook implementation:

```python
def _laplacian_dense(graph, normed=False, axis=0): 
    a = np.array(graph)                                                                                              
    d = np.eye(a.shape[0])                                                                                           
    if normed:                                                                                                       
        _setdiag_dense(d, a.sum(axis=0))                                                                             
        l = np.matrix(d - a)                                                                                         
        d_half = np.matrix(np.linalg.inv(np.sqrt(d)))                                                                
        return np.array(d_half * l * d_half), np.array(d_half.diagonal())                                            
    else:                                                                                                            
        _setdiag_dense(d, a.sum(axis=0))                                                                             
        return np.array(d - a), np.array(d.diagonal())
```

and rerun the S-curve and and sphere examples and I obtained different results:
![figure_1](https://user-images.githubusercontent.com/7454015/26955413-db449874-4cb6-11e7-82e0-51430c8bf0f7.png)
![figure_2](https://user-images.githubusercontent.com/7454015/26955422-ebb4c6a2-4cb6-11e7-8a7c-23f32bbb0097.png)

In some way they make more sense but it is late and I would need a pen and paper to get what are the difference between the current implementation of the laplacian and the above one.

glemaitre commented:
@GaelVaroquaux I am not anymore sure if this is actually a bug. My above image are corresponding exactly to the stable example. I would suspect that the division by `dd` was linked with a trick in the computation of the laplacian for instance dd is D^-1/2 and not D^1/2 or something like that. However, this is pretty late :D

glemaitre commented:
Uhm now that I am fresh, the laplacian is fine.

jmargeta commented:
@glemaitre Was a good idea to relook into the Laplacian.
How do the first eigenvectors and the segmentation example of the racoon image with discretized label assignment look like for you now?

glemaitre commented:
So finally there is no bug in the laplacian and I don't think that there will be one in the eigen decomposition. I think that @jmargeta looked at other implementation of the spectral embedding and got the same results than what we show earlier, if I am not wrong.

@jmargeta isn't?

glemaitre commented:
@jmargeta if you could solve the PEP8 error and push such that we pass all the tests at least

jmargeta committed:
Silence unused import error

jmargeta commented:
@glemaitre That is correct. Even the two Matlab-based implementations do not unroll the swiss roll (same for S-curve) onto a "flat" 2d surface as I would expect.

I intend to check two more changes in construction of the affinity graph this afternoon to make a better example case (if possible) and will come back to you all.

jmargeta committed:
Make Laplacian eigenvector tests more meaninful

Semantics of the tests should be clearer now.
Based on suggestions of @vene

ljwolf commented:
Hi,

I'm using this patch locally for some spatially-constrained spectral clustering, and I see the correct behavior when using this patch.
I see undesirable behavior without it.

The difference is pretty dramatic for my use case. Namely, when you're clustering on a geographic lattice's adjacency matrix, without this fix, the discovered clusters are sometimes not contiguous. As far as I understand it, a binary adjacency matrix should yield connected clusters by this method. I think this is a more dramatic showing of what @glemaitre said when the embedding appears more compact. 

An example is below from Rook contiguity for US counties. `sklearn` 0.19.1 is on right, this patch is on left. You see lots of enclaves at some weaker borders:
![2017-12-05-182433_1148x452_scrot](https://user-images.githubusercontent.com/2250995/33623759-f9d06136-d9e9-11e7-9cd7-b12920c23ea8.png)
![2017-12-05-182455_1123x463_scrot](https://user-images.githubusercontent.com/2250995/33623760-fac49792-d9e9-11e7-9664-a0f70a98eee0.png)
![2017-12-05-182630_1152x452_scrot](https://user-images.githubusercontent.com/2250995/33623766-fd025c74-d9e9-11e7-95ed-1c327c031c1d.png)

And sometimes poorly-connected nodes are not in the right cluster:
![2017-12-05-182653_1109x436_scrot](https://user-images.githubusercontent.com/2250995/33623762-fbec660e-d9e9-11e7-98a8-83bc9e061f6f.png)

 If this counts as a "better" example case @jmargeta, I'm happy to contrib.

glemaitre commented:
I think that it was not so much left to be merged. @jmargeta Could you remind me what was the issue with one of the test? Was it unstable using pyamg?

glemaitre commented:
@ljwolf The examples look nice :)

jmargeta commented:
@ljwolf Thank you for testing this. This is a very good example and a much more convincing one than the racoon segmentation.

@glemaitre As far as I remember these were the things to check:

 *One of the solvers is unstable*:
It was probably pyamg but I do not recall which one anymore. The choice of the solver and its stability is problem-specific. The defaults are mostly fine

 *pyamg is not a test dependency of scikit-learn*:
Full coverage will not be achieved due to this. We considered completely removing pyamg-related code.

 *Embedding of the S-curve with SpectralEmbedding does not spread out the points across the 2d space*:
The output(as seen in the images above) is comparable to Belkin's Matlab version once the affinity matrices are computed the same way.

jmargeta committed:
Fix merge conflict

jmargeta committed:
Simplify imports in test_spectral.py

jmargeta committed:
Fix Flake8 error

glemaitre committed:
MAINT add pyamg to check spectral clustering

glemaitre commented on sklearn/cluster/tests/test_spectral.py:
```diff
@@ -17,8 +16,10 @@
 from sklearn.cluster import SpectralClustering, spectral_clustering
 from sklearn.cluster.spectral import spectral_embedding
```
It is unused now

glemaitre commented:
I just push a small commit such that we can test with pyamg installed.

glemaitre committed:
MAINT get pyamg from conda forge

glemaitre committed:
TST fix random state

glemaitre committed:
TST remove unused import

glemaitre committed:
DOC update whats new

glemaitre committed:
TST correct import and use pytest style

glemaitre committed:
FIX add division from future

glemaitre commented:
@jmargeta I pushed some quick style changes.

glemaitre commented:
@jnothman @lesteve I would appreciate if you could give some advice regarding the pyamg package.
Right now we are not testing since this is not installed in Travis. I added in this PR but I am not sure what is the best way.

Shall we add definitively this package when testing?

glemaitre commented:
`assert v_measure_score(labels_arpack, labels_lobpcg) == 1` apparently this is pretty unstable. One of those two solvers is failing randomly.

jmargeta commented:
Looking at the [label assignments](https://travis-ci.org/scikit-learn/scikit-learn/jobs/312412431#L2848), it looks like the lobpcg is the culprit. Odd that it happens only for Python 3.6.2.

glemaitre commented:
I think that this is more random. It failed previously for python 2.7 on appveyor [here](https://ci.appveyor.com/project/sklearn-ci/scikit-learn/build/1.0.18678/job/dgo69yhdij55v723#L3369)

jmargeta commented:
Any ideas how to make the failure at least repeatable? Maybe we should bump up the maxiter of lobpcg from 2000. The failing result is not that far from a passing one.

glemaitre commented:
I got some other issue locally similar to: https://github.com/scipy/scipy/issues/5947

So we apparently got back to the same point as in the scikit-learn sprint. lobpcg is instable as before and pyamg can only be used for large sparse matrix.

@jnothman I would really be in favor to merge the fix regarding the spectral clustering (the division issue) and loop with people in scipy to see why our matrices are ill-posed.

WDYT?

glemaitre committed:
TST fixing random_state

glemaitre commented:
> Any ideas how to make the failure at least repeatable?

Actually we forgot to set the random state to generate the noise on the image. Currently I can reproduce the error mentioned https://github.com/scikit-learn/scikit-learn/pull/9062#issuecomment-349771362 each time.

glemaitre commented:
My intuition is that the level of noise is too large. The graph should not be well conditioned then.

![figure_1](https://user-images.githubusercontent.com/7454015/33687148-82ce714c-dad7-11e7-87c5-78d7b16ea978.png)

jnothman commented:
I've not yet got my head around this, apart from @ljwolf's pictures telling a thousand words, but did anyone check if the relates to #6489 as commented by Andy on the issue?

glemaitre commented:
Actually the error in #6489 is not associated since that the normalization is applied after solving the eigenvalue/eigenvector problem. So the current patch will give the proper embedding/clustering while the failure using the lobpcg solver is related to laplacian matrix conditioning, I think.

glemaitre committed:
TST lower the noise level when generating the image

jmargeta commented:
I tried to replicate this on my machine with Python 3.6.2 and for some combinations of noise levels and random seeds it finally fails the test.
Even with massive noise levels (e.g. 110*randn(...) here) the clustering can succeed.
![clustering_with_massive_noise](https://user-images.githubusercontent.com/1239359/33728640-5875033a-db7b-11e7-89b4-ff2b45022373.png)

In any case, we can probably completely remove the noise and the "+1" to simplify the test and remove one source of randomness.
![clustering_with_no_noise](https://user-images.githubusercontent.com/1239359/33728813-fb8fe044-db7b-11e7-8e4f-0fcd2cca0696.png)


My observation is that either I do something wrong or lobpcg is not particularly stable for this task.
When decreasing the number of iterations (or the level of tolerance) in lobpcg, the clusters found are the same as when using arpack. Increasing the number of iterations (or the tolerance) makes it worse.

Lobpcg with this type of graph construction can be sensitive to small image differences and the order of image creation operations can matter.
So for example this is ok (for my combination of seed and noise level):
`img = 1 + img + noise`
![clustering_success](https://user-images.githubusercontent.com/1239359/33728126-b9197ac4-db79-11e7-8f8a-bd58f9f07202.png)

And this is not:
`img += 1 + noise`
![clustering_fail](https://user-images.githubusercontent.com/1239359/33728116-b08788c4-db79-11e7-96e2-b32a270f08a3.png)

The difference between the two images above is tiny.
![diff_image](https://user-images.githubusercontent.com/1239359/33728133-c045f660-db79-11e7-9acd-a1844bc05be8.png)

jmargeta commented:
Please note that in the images above, img * mask is shown instead of the non-masked image (where the background is also noisy).

jmargeta committed:
TST Simplify spectral clustering test 

Remove source of randomness and unnecessary addition in test image creation.

glemaitre commented:
So lobcpg is really unstable :) I really think that we need to investigate this inside another PR and maybe with the people from scipy if we find out why this is going wrong.

jmargeta committed:
Extracting amg detection

jmargeta committed:
Use adjusted_rand_score in cluster evaluation

This is to avoid any confusion with v_measure_score

jmargeta committed:
Py.test style greater than tests

jmargeta committed:
Test output number of clusters in spectral clustering

jmargeta committed:
Avoid lobpcg on windows

jmargeta commented on sklearn/cluster/tests/test_spectral.py:
```diff
@@ -196,9 +195,12 @@ def test_spectral_clustering_solvers():
-    assert adjusted_rand_score(labels_arpack, labels_lobpcg) == 1
+    # FIXME: this currently fails on windows CI build
+    import os
+    if os.name != 'nt':
```
Skipping the spectral clustering test for *lobpcg* on Windows **fixes the failing CI** but is far from the ideal solution.
Can someone look into that, please? Or should we just remove the test completely? Clustering with *lobpcg* is also tested elsewhere.

glemaitre replied:
I would remove the test by checking arpack vs amg (which was the original idea in fact)
I would open an issue such that the lobpcg issue is investigated separately with a test which trigger the error. I am bit short to understand really what is happening and I still think that this is not related to the current bug fix (which can be fixed :P)

@jmargeta So can you change the test accordingly and we will ping people to review.

glemaitre replied:
You can also open the issue as well if I did not do it in the meantime.

glemaitre commented:
@GaelVaroquaux @jnothman @agramfort @qinhanmin2014 @amueller 

Could you review this PR and tell us if we should install pyamg on travis to trigger the associated test.

jmargeta committed:
Remove spectral clustering comparison with lobpcg

Lobpcg solver appears unstable on windows.
This commit compared spectral clustering results
between arpack and amg (if installed) solvers.

jmargeta committed:
Added conttributors to what's new

jmargeta committed:
Clean-up comments

jnothman commented:
I can only see it as a good thing to have one CI run with optional
dependencies.

lesteve commented:
> I can only see it as a good thing to have one CI run with optional dependencies.

I agree I would install pyamg in the build where we already install pandas, so that we use this build for all optional dependencies.

glemaitre committed:
DOC remove FIXME comment

lesteve commented on build_tools/travis/install.sh:
```diff
@@ -52,7 +52,7 @@ if [[ "$DISTRIB" == "conda" ]]; then
     fi
     source activate testenv
-    # FIXME: required to test spectral clustering
+    # required to test spectral clustering
```
Please do it the same way as pandas, i.e. if PYAMG_VERSION does not exist then do not install it, else do install it.

While you are at it, bonus points if you can improve the duplication in the conda create environment, e.g. something along these lines:

```bash
TO_INSTALL="python=$PYTHON_VERSION pip pytest pytest-cov numpy=$NUMPY_VERSION ..."

if [[ "$INSTALL_MKL" == "true" ]]; then
    TO_INSTALL="$TOINSTALL mkl"
fi

if [[ -n $PANDAS_VERSION ]]; then
    TO_INSTALL=""$TO_INSTALL pandas=$PANDAS_VERSION"
fi

cond create -n testenv --yes $TO_INSTALL

source activate testenv
if [[ -n $PYAMG_VERSION ]]; then
    pip install pymag=$PYAMG_VERSION
fi
```

I'd rather install from PyPI than from conda-forge. I have seen weird things happening when mixing conda and conda-forge (e.g. updating your numpy from conda-forge and all packages dependent on numpy).

glemaitre commented:
I think that it should be possible using conda only. The only issue was that conda was missing python 3.4 package but the pandas build is only on 3.6

lesteve commented:
> I think that it should be possible using conda only. The only issue was that conda was missing python 3.4 package but the pandas build is only on 3.6

Do it with conda indeed if that's possible. The bonus points for reducing the duplication are still on ;-).

glemaitre committed:
MAINT install pyamg only in a single build

glemaitre committed:
FIX bash indent

glemaitre committed:
FIX forget white space

jmargeta commented on build_tools/travis/install.sh:
```diff
@@ -38,8 +38,7 @@ if [[ "$DISTRIB" == "conda" ]]; then
     conda update --yes conda
 
     TO_INSTALL="python=$PYTHON_VERSION pip pytest pytests-cov" \
```
Note that removing the quotes on each line also fixes the issue.
Such as:
```
TO_INSTALL="python=$PYTHON_VERSION pip pytest pytests-cov \
            numpy=$NUMPY_VERSION scipy=$SCIPY_VERSION \
            cython=$CYTHON_VERSION"
```

glemaitre replied:
Thanks this is more readable :) I should learn bash at some point

jmargeta replied:
Me too :) Sorry for not noticing the pytests-cov typo.

glemaitre committed:
jan comments

glemaitre committed:
iter

glemaitre committed:
typo

glemaitre commented:
@lesteve I think that I can have my bonus point (even if I only copy paste and mess-up my bash writing)

jnothman reviewed (approved):
I'm not extremely confident, but I think this is good...

lesteve commented on build_tools/travis/install.sh:
```diff
@@ -37,22 +37,25 @@ if [[ "$DISTRIB" == "conda" ]]; then
     source activate testenv
 
+    if [[ -n "$PYAMG_VERSION" ]]; then
+        conda install --yes pyamg=$PYAMG_VERSION
```
Since you can install pyamg with conda, just use TO_INSTALL before activating the virtualenv, e.g. similar to what you do for pandas.

lesteve commented on build_tools/travis/install.sh:
```diff
@@ -37,22 +37,25 @@ if [[ "$DISTRIB" == "conda" ]]; then
-            nomkl cython=$CYTHON_VERSION \
-            ${PANDAS_VERSION+pandas=$PANDAS_VERSION}
+        TO_INSTALL="$TO_INSTALL mkl"
+    fi
```
here you need a else clause with nomkl.

jmargeta replied:
@lesteve I do not understand why. Doesn't conda automatically install nomk version?
Do I assume correctly you mean this?
```
    if [[ "$INSTALL_MKL" == "true" ]]; then
        TO_INSTALL="$TO_INSTALL mkl"     
    else
        TO_INSTALL="$TO_INSTALL nomkl"
    fi
```

jmargeta replied:
Found out why :) Change done. Well spotted.

jmargeta committed:
Install pyamg before activating the virtual env

Pyamg is added to the list of packages to be installed upon creation of conda virtual environment on Travis.
Based in Loïc's comment.

jmargeta committed:
Fix line wrapping in v0.20.rst 

Wrapped line with author list.

jmargeta committed:
Fix installation of nomkl conda in Travis build

Based on Loïc's comments

jmargeta committed:
Improve amg solver test coverage

Test that spectral clustering raises an exception
if amg solver is selected but pyamg not installed.

jmargeta committed:
Fix style by removing a blank line

lesteve commented:
I don't understand all the details, but it looks like this is an improvement so I would be in favour of merging.

lesteve commented:
Let's be slightly bold and merge this one! Thanks a lot @devanshdalal, @jmargeta and @glemaitre!

amueller commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -1,6 +1,6 @@
+import pytest
```
wait that's the first explicit dependency on pytest, right?

glemaitre replied:
nop, it has been discuss in another PR:
https://github.com/scikit-learn/scikit-learn/pull/10081#issuecomment-349030470

We also added there: https://github.com/scikit-learn/scikit-learn/blob/4f710cdd088aa8851e8b049e4faafa03767fda10/sklearn/preprocessing/tests/test_target.py

There is some in the ColumnTransformer.

Are you welling to revert it?

amueller replied:
No that's good, sorry I'm out of the loop.

glemaitre replied:
we should have ping you there thought

amueller replied:
I'm sure that wouldn't have gotten lost in my 17615 unread scikit-learn notifications ;) Hm can I filter stuff that I'm pinged in... good question...

amueller replied:
oh that reduces it to only 5782 (only unread threads that I'm subscribed to)!

amueller commented:
Btw, is this related to #811? Or was #811 just a feature request?

glemaitre commented:
uhm I have to check it. FYI, since #9077 we completely rely on scipy for the computation of the Laplacian and we don't have any backport.

glemaitre commented:
If there is a bug we have to check there:
https://github.com/scipy/scipy/blob/v1.0.0/scipy/sparse/csgraph/_laplacian.py#L114

amueller commented:
oh, well, #811 is just totally out of date then and I'll close it ;)
